celo-org/celo-blockchain
# Hotswap Validators

### Description

This PR add "replica" validators which can start mining at a given block. Primary validators can be instructed to stop at a given block through the RPC API.

Validators can be started as replicas with the `--istanbul.replica` flag. They should still be provided `--mine` because they start "mining" without participating in consensus. This is done to allow the node to participate in a limited form of announce to get the enode urls of the other validators.

To switch validators
-  spin up a replica
- Once synced & it has the enode urls for the rest of the validators, pick a block in the future to switch on.
- Run `istanbul.stopAtBlock(BLOCK_NUM)` on the primary
- Run `istanbul.startAtBlock(BLOCK_NUM)` on the replica
- You can verify the switch block in `istanbul.replicaState`. Note the switch block number should be the same for the primary and replica.
- spin down the old primary

Added RPC Calls
* `istanbul.start()` Immediately make a replica the primary & participate in consensus.
* `istanbul.stop()` Immediately make a primary a replica & stop participating consensus.
* `istanbul.startAtBlock(blockNum)` The node will start validating on the given block # 
* `istanbul.stopAtBlock(blockNum)` The node will stop validating on the block number. (Will not validate the stop block)
* `istanbul.validating` If this node is actively validating
* `istanbul.replicaState ` Information on the block that this node will switch at and if this is a primary or replica


**To Do for testing**:
- Add test for killing validator and restarting it to check that it persists replica state.


### Other changes

Did rework some of announce & the backend message handler for clarity.

### Tested



Tested manually the following flows:
- Start network, spin up secondary, switch to secondary, switch back to primary.
- Start network, spin up secondary, schedule start & stop, cancel start & stop, switch to secondary
- Start network, spin up secondary, use start/stop API to switch to and from primary (will skip blocks b/c manual)

Unit tests - for `replicaState` and `checkMessage`
e2e tests - done, see  https://github.com/celo-org/celo-monorepo/pull/5029

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

Yes
